File: TP1/test_encapsulation.py
import pytest
from src.personne import Personne
from src.etudiant import Etudiant

def test_validation_age():
    p = Personne("Alice", 25)
    with pytest.raises(ValueError):
        p.age = -5
    with pytest.raises(ValueError):
        p.age = 105

def test_validation_moyenne():
    e = Etudiant("Bob", 20, "ETU001", 10.0)
    with pytest.raises(ValueError):
        e.moyenne = 25.0

def test_validation_nom_vide():
    with pytest.raises(ValueError):
        Personne("", 20)

File: TP1/java/TestEncapsulation.java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class TestEncapsulation {
    @Test
    void testAgeInvalide() {
        Personne p = new Personne("Alice", 25);
        assertThrows(IllegalArgumentException.class, () -> p.setAge(-1));
        assertThrows(IllegalArgumentException.class, () -> p.setAge(101));
    }

    @Test
    void testMoyenneInvalide() {
        Etudiant e = new Etudiant("Bob", 20, "ETU001", 10.0);
        assertThrows(IllegalArgumentException.class, () -> e.setMoyenne(22.0));
    }

    @Test
    void testNomVide() {
        assertThrows(IllegalArgumentException.class, () -> new Personne("", 20));
    }
}
